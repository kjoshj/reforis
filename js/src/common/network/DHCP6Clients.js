/*
 * Copyright (C) 2019-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { formFieldsSize, RichTable } from "foris";
import moment from "moment";
import PropTypes from "prop-types";

DHCP6Clients.propTypes = {
    ipv6clients: PropTypes.arrayOf(PropTypes.object),
};

DHCP6Clients.defaultProps = {
    ipv6clients: [],
};

export default function DHCP6Clients({ ipv6clients }) {
    function renderActiveIcon({ getValue }) {
        return <ActiveIcon isActive={getValue()} />;
    }

    const columns = [
        {
            accessorKey: "hostname",
            header: _("Hostname"),
        },
        {
            accessorKey: "ipv6",
            header: _("IPv6 Address"),
        },
        {
            accessorKey: "duid",
            header: _("DUID"),
        },
        {
            accessorKey: "expires",
            header: _("Expires"),
            cell: ({ getValue }) => {
                const expires = getValue();
                return moment
                    .utc(expires)
                    .locale(ForisTranslations.locale)
                    .format("l LT");
            },
        },
        {
            accessorKey: "active",
            header: _("Active"),
            headerClassName: "d-flex justify-content-center",
            className: "text-center",
            cell: renderActiveIcon,
        },
    ];

    const data = ipv6clients.map((client) => ({
        hostname: client.hostname,
        ipv6: client.ipv6,
        duid: client.duid,
        active: client.active,
        expires: client.expires
            ? moment
                  .unix(client.expires)
                  .locale(ForisTranslations.locale)
                  .format("l LT")
            : _("Never"),
    }));

    return (
        <div className={formFieldsSize}>
            <h2>{_("IPv6 DHCP Client List")}</h2>
            <p>
                {_(
                    "This list contains all devices that are connected to the network through wired or wireless connections using IPv6."
                )}
            </p>
            {ipv6clients.length === 0 ? (
                <p className="text-muted text-center">
                    {_("No clients found.")}
                </p>
            ) : (
                <RichTable columns={columns} data={data} withPagination />
            )}
        </div>
    );
}

ActiveIcon.propTypes = {
    isActive: PropTypes.bool,
};

function ActiveIcon({ isActive }) {
    if (isActive) {
        return (
            <FontAwesomeIcon
                icon="fa-check"
                className="text-success text-center"
                title={_("Device is active")}
            />
        );
    }
    return (
        <FontAwesomeIcon
            icon="fa-times"
            className="text-danger"
            title={_("Device is inactive")}
        />
    );
}
