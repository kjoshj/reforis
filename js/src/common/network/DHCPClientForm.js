/*
 * Copyright (C) 2019-2022 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import { TextInput, validateHostname } from "foris";
import PropTypes from "prop-types";

const HELP_TEXTS = {
    hostname: _("Hostname which will be provided to DHCP server."),
};

DHCPClientForm.propTypes = {
    formData: PropTypes.shape({
        hostname: PropTypes.string,
    }).isRequired,
    formErrors: PropTypes.shape({
        hostname: PropTypes.string,
    }),
    setFormValue: PropTypes.func.isRequired,
    updateRule: PropTypes.func.isRequired,
    disabled: PropTypes.bool,
};

export default function DHCPClientForm({
    formData,
    formErrors,
    setFormValue,
    updateRule,
    disabled,
}) {
    return (
        <TextInput
            label={_("DHCP hostname")}
            value={formData.hostname || ""}
            error={(formErrors || {}).hostname || null}
            helpText={HELP_TEXTS.hostname}
            onChange={setFormValue((value) =>
                updateRule({ hostname: { $set: value } })
            )}
            disabled={disabled}
        />
    );
}

export function validateDHCPForm(formData) {
    const error = { hostname: validateHostname(formData.hostname) };
    return error.hostname ? error : null;
}
