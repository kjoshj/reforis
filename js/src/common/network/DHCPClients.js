/*
 * Copyright (C) 2019-2025 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import { formFieldsSize } from "foris";
import PropTypes from "prop-types";

import DHCPClientsTable from "common/network/DHCPClients/DHCPClientsTable";
import DHCPClientsWithStaticLeases from "common/network/DHCPClients/DHCPClientsWithStaticLeases";

DHCPClients.propTypes = {
    clients: PropTypes.arrayOf(PropTypes.object),
    withStaticLeases: PropTypes.bool,
};

export default function DHCPClients({ clients, withStaticLeases }) {
    return (
        <div className={formFieldsSize}>
            <h2>{_("DHCP Client List")}</h2>
            <p>
                {_(
                    "This list contains all devices that are connected to the network through wired or wireless connections."
                )}
            </p>
            {withStaticLeases ? (
                <DHCPClientsWithStaticLeases />
            ) : (
                <DHCPClientsTable clients={clients} />
            )}
        </div>
    );
}
