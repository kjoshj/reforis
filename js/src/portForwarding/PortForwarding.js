/*
 * Copyright (C) 2019-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useState, useEffect, useCallback } from "react";

import {
    Button,
    withErrorMessage,
    withSpinnerOnSending,
    formFieldsSize,
    ForisURLs,
} from "foris";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";

import useRulesForm, {
    usePortForwarding,
    usePortForwardingWS,
    usePortForwardingRuleDelete,
} from "portForwarding/hooks";
import RulesModal from "portForwarding/PortForwardingRulesModal";
import RulesTable from "portForwarding/PortForwardingRulesTable";

PortForwarding.propTypes = {
    ws: PropTypes.object,
};

export default function PortForwarding({ ws }) {
    const {
        lanDetails,
        portForwardingRules,
        lanState,
        portForwardingRulesState,
        getPortForwardingRules,
        setRules,
    } = usePortForwarding();

    const staticLeases =
        lanDetails.mode_managed?.dhcp?.clients?.filter(
            (client) => client.static
        ) || [];

    return (
        <>
            <h1>{_("Port Forwarding")}</h1>
            <p className="mb-0">
                {_(
                    "Port forwarding allows you to forward incoming traffic from the WAN interface to a specific device on your LAN."
                )}
            </p>
            <p>
                {_(
                    "For example, you can forward incoming traffic on port 80 to a web server on your LAN. This way, the web server will be accessible from the internet."
                )}
            </p>
            <p>
                {_(
                    'You can add a new port forwarding rule by clicking the "Add rule" button. But first, you need to configure a static DHCP lease for the device you want to forward traffic to. You can configure it on the '
                )}
                {_("")}
                <Link to={ForisURLs.lanSettings}>{_("LAN")}</Link>
                {_(" page.")}
            </p>
            <PortForwardingWithErrorAndSpinner
                apiState={[portForwardingRulesState.state, lanState.state]}
                rulesList={portForwardingRules}
                staticLeases={staticLeases}
                refetchPortForwardingRules={getPortForwardingRules}
                setRules={setRules}
                ws={ws}
            />
        </>
    );
}

PortForwardingConfigurator.propTypes = {
    rulesList: PropTypes.array,
    staticLeases: PropTypes.array,
    refetchPortForwardingRules: PropTypes.func,
    setRules: PropTypes.func,
    ws: PropTypes.object,
};

function PortForwardingConfigurator({
    rulesList,
    staticLeases,
    refetchPortForwardingRules,
    setRules,
    ws,
}) {
    const [ruleToEdit, setRuleToEdit] = useState(null);
    const [ruleModalShown, setRuleModalShown] = useState(false);
    const [ruleToBeDeleted, setRuleToBeDeleted] = useState(null);

    const { deleteRule } = usePortForwardingRuleDelete(
        ruleToBeDeleted,
        refetchPortForwardingRules
    );

    const postCallback = useCallback(() => {
        setRuleModalShown(false);
        refetchPortForwardingRules();
    }, [setRuleModalShown, refetchPortForwardingRules]);

    const [
        formState,
        setFormValue,
        postNewRuleState,
        postOldRuleState,
        saveForwardingRule,
    ] = useRulesForm(ruleToEdit, postCallback, setRuleModalShown);

    usePortForwardingWS(ws, setRules, postNewRuleState, postOldRuleState);

    const handleAddRule = () => {
        setRuleToEdit(null);
        setRuleModalShown(true);
    };

    const handleEditRule = (rule) => {
        setRuleToEdit(rule);
        setRuleModalShown(true);
    };

    const handleDeleteRule = (name) => {
        setRuleToBeDeleted(name);
    };

    useEffect(() => {
        if (ruleToBeDeleted) {
            deleteRule();
        }
    }, [ruleToBeDeleted, deleteRule]);

    const ruleModalTitle = ruleToEdit
        ? _("Edit port forwarding rule")
        : _("Add port forwarding rule");

    return (
        <div className={formFieldsSize}>
            <RulesTable
                rules={rulesList}
                onEditRule={handleEditRule}
                onDeleteRule={handleDeleteRule}
            />
            <AddRuleButton addRule={handleAddRule} />
            <RulesModal
                shown={ruleModalShown}
                setShown={setRuleModalShown}
                title={ruleModalTitle}
                rule={ruleToEdit}
                staticLeases={staticLeases}
                formState={formState}
                setFormValue={setFormValue}
                saveForwardingRule={saveForwardingRule}
                postOldRuleState={postOldRuleState}
                postNewRuleState={postNewRuleState}
            />
        </div>
    );
}

const PortForwardingWithErrorAndSpinner = withSpinnerOnSending(
    withErrorMessage(PortForwardingConfigurator)
);

AddRuleButton.propTypes = {
    addRule: PropTypes.func,
};

function AddRuleButton({ addRule }) {
    return (
        <div className="text-end">
            <Button
                forisFormSize
                className="btn-primary btn mb-3 mb-sm-2 mb-md-0"
                onClick={addRule}
            >
                {_("Add rule")}
            </Button>
        </div>
    );
}
