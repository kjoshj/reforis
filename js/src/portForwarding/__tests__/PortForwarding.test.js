/*
 * Copyright (C) 2019-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import { WebSockets } from "foris";
import { fireEvent, render, wait } from "foris/testUtils/customTestRender";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";

import diffSnapshot from "snapshot-diff";
import mockAxios from "jest-mock-axios";

import PortForwarding from "portForwarding/PortForwarding";
import lanSettingsFixture from "portForwarding/__fixtures__/lanSettings";
import portForwardingFixture from "portForwarding/__fixtures__/portForwarding";

describe("<PortForwarding/>", () => {
    let firstRender;
    let asFragment;
    let getByText;
    let getByLabelText;
    let getByTestId;

    beforeEach(async () => {
        const webSockets = new WebSockets();
        const renderRes = render(
            <>
                <PortForwarding ws={webSockets} />
                <div id="modal-container" />
            </>
        );
        getByText = renderRes.getByText;
        getByLabelText = renderRes.getByLabelText;
        getByTestId = renderRes.getByTestId;
        asFragment = renderRes.asFragment;

        mockAxios.mockResponse({ data: portForwardingFixture });
        mockAxios.mockResponse({ data: lanSettingsFixture });

        await wait(() => renderRes.getByText("Add rule"));
        firstRender = renderRes.asFragment();
    });

    it("should handle error when fetching data fails", async () => {
        const webSockets = new WebSockets();
        const { getByText } = render(<PortForwarding ws={webSockets} />);

        mockJSONError();
        mockJSONError();

        await wait(() => {
            expect(
                getByText("An error occurred while fetching data.")
            ).toBeTruthy();
        });
    });

    it("should render page with one rule defined", () => {
        expect(firstRender).toMatchSnapshot();
    });

    it("should open modal when adding rule", () => {
        const addRuleButton = getByText("Add rule");
        addRuleButton.parentElement.click();
        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it("should add port forwarding rule", () => {
        const addRuleButton = getByText("Add rule");
        addRuleButton.parentElement.click();

        fireEvent.change(getByLabelText("Name"), {
            target: { value: "Test" },
        });
        fireEvent.change(getByLabelText("Internal IPv4 address"), {
            target: { value: "192.168.2.102" },
        });
        fireEvent.change(getByLabelText("External port"), {
            target: { value: "8090" },
        });
        fireEvent.change(getByLabelText("Internal port"), {
            target: { value: "8090" },
        });
        fireEvent.click(getByLabelText("Enabled"));

        const saveButton = getByText("Save");
        saveButton.parentElement.click();

        expect(mockAxios.post).toBeCalled();

        const data = {
            name: "Test",
            dest_ip: "192.168.2.102",
            dest_port: "8090",
            src_dport: "8090",
            enabled: true,
        };
        expect(mockAxios.post).toHaveBeenCalledWith(
            "/reforis/api/lan/port-forwarding",
            data,
            expect.anything()
        );
    });

    it("should delete rule", async () => {
        expect(getByText("Service")).toBeTruthy();
        const threeDotsDelBtn = getByTestId("three-dots-menu-delete");

        fireEvent.click(threeDotsDelBtn);

        expect(mockAxios.delete).toBeCalled();
        expect(mockAxios.delete).toHaveBeenCalledWith(
            "/reforis/api/lan/port-forwarding/Service",
            expect.anything()
        );
    });

    it("should edit rule", async () => {
        const threeDotsEditBtn = getByTestId("three-dots-menu-edit");
        fireEvent.click(threeDotsEditBtn);

        fireEvent.change(getByLabelText("Name"), {
            target: { value: "Test" },
        });

        const saveButton = getByText("Save");
        saveButton.parentElement.click();

        expect(mockAxios.post).toBeCalled();
        expect(mockAxios.post).toHaveBeenCalledWith(
            "/reforis/api/lan/port-forwarding",
            {
                old_name: "Service",
                name: "Test",
                dest_ip: "192.168.2.101",
                dest_port: 8090,
                src_dport: 8090,
                enabled: true,
            },
            expect.anything()
        );
    });

    it("should handle error when adding rule fails: port is already in use", async () => {
        expect(getByText("Service")).toBeTruthy();

        const addRuleButton = getByText("Add rule");
        addRuleButton.parentElement.click();

        fireEvent.change(getByLabelText("Name"), {
            target: { value: "Test" },
        });
        fireEvent.change(getByLabelText("Internal IPv4 address"), {
            target: { value: "192.168.2.102" },
        });
        fireEvent.change(getByLabelText("External port"), {
            target: { value: "8090" },
        });
        fireEvent.change(getByLabelText("Internal port"), {
            target: { value: "8090" },
        });
        fireEvent.click(getByLabelText("Enabled"));

        const saveButton = getByText("Save");
        saveButton.parentElement.click();

        mockJSONError("Port is already in use.");

        await wait(() => {
            expect(mockSetAlert).toBeCalledWith("Port is already in use.");
        });
    });
});
