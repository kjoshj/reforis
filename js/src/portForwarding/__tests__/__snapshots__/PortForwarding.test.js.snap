// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PortForwarding/> should open modal when adding rule 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -160,11 +160,11 @@
      id=\\"modal-container\\"
    >
      <div
        aria-labelledby=\\"modal-title\\"
        aria-modal=\\"true\\"
-       class=\\"modal fade\\"
+       class=\\"modal fade show\\"
        role=\\"dialog\\"
      >
        <div
          class=\\"modal-dialog modal-dialog-centered modal-dialog-scrollable\\"
          role=\\"document\\""
`;

exports[`<PortForwarding/> should render page with one rule defined 1`] = `
<DocumentFragment>
  <h1>
    Port Forwarding
  </h1>
  <p
    class="mb-0"
  >
    Port forwarding allows you to forward incoming traffic from the WAN interface to a specific device on your LAN.
  </p>
  <p>
    For example, you can forward incoming traffic on port 80 to a web server on your LAN. This way, the web server will be accessible from the internet.
  </p>
  <p>
    You can add a new port forwarding rule by clicking the "Add rule" button. But first, you need to configure a static DHCP lease for the device you want to forward traffic to. You can configure it on the 
    <a
      href="/network-settings/lan"
    >
      LAN
    </a>
     page.
  </p>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Port Forwarding Rules
    </h2>
    <div
      class="table-responsive"
    >
      <table
        class="table table-hover text-nowrap"
      >
        <thead
          class="thead-light"
        >
          <tr>
            <th
              scope="col"
            >
              Name
            </th>
            <th
              scope="col"
            >
              Internal IP address
            </th>
            <th
              scope="col"
            >
              External port
            </th>
            <th
              scope="col"
            >
              Internal port
            </th>
            <th
              class="text-center"
              scope="col"
            >
              Enabled
            </th>
            <th
              aria-label="Actions"
              scope="col"
            />
          </tr>
        </thead>
        <tbody>
          <tr
            class="align-middle"
          >
            <td>
              Service
            </td>
            <td>
              192.168.2.101
            </td>
            <td>
              8090
            </td>
            <td>
              8090
            </td>
            <td
              class="text-center"
            >
              <i
                class="fa"
              />
            </td>
            <td
              class="text-end"
            >
              <div
                class="dropdown position-static"
                data-testid="three-dots-menu"
              >
                <button
                  aria-expanded="false"
                  class="btn btn-sm btn-link text-body d-inline-flex justify-content-center align-items-center"
                  data-bs-toggle="dropdown"
                  type="button"
                >
                  <span>
                    <i
                      class="fa"
                    />
                  </span>
                </button>
                <ul
                  class="dropdown-menu"
                >
                  <li>
                    <button
                      class="dropdown-item"
                      data-testid="three-dots-menu-edit"
                      type="button"
                    >
                      <i
                        class="fa"
                      />
                      Edit
                    </button>
                  </li>
                  <li>
                    <button
                      class="dropdown-item"
                      data-testid="three-dots-menu-delete"
                      type="button"
                    >
                      <i
                        class="fa"
                      />
                      Delete
                    </button>
                  </li>
                </ul>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="text-end"
    >
      <button
        class="btn btn-primary btn mb-3 mb-sm-2 mb-md-0 col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
        type="button"
      >
        <span>
          Add rule
        </span>
      </button>
    </div>
  </div>
  <div
    id="modal-container"
  >
    <div
      aria-labelledby="modal-title"
      aria-modal="true"
      class="modal fade"
      role="dialog"
    >
      <div
        class="modal-dialog modal-dialog-centered modal-dialog-scrollable"
        role="document"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h1
              class="modal-title fs-5"
            >
              Add port forwarding rule
            </h1>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body"
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="1"
              >
                Name
              </label>
              <div
                class="input-group"
              >
                <input
                  class="form-control"
                  id="1"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-text"
              >
                <small>
                  Enter the name of the port forwarding rule.
                </small>
              </div>
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="2"
              >
                Internal IPv4 address
              </label>
              <div
                class="input-group"
              >
                <input
                  class="form-control rounded"
                  id="2"
                  list="addresses"
                  type="text"
                  value=""
                />
                <datalist
                  aria-label="List of available IPv4 addresses"
                  id="addresses"
                >
                  <option
                    label="MacBook"
                    value="192.168.2.101"
                  >
                    MacBook
                  </option>
                </datalist>
              </div>
              <div
                class="form-text"
              >
                <small>
                  Enter a valid IP address of the device to which the port will be forwarded. The best practice is to assign a static lease to the device in advance.
                </small>
              </div>
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="3"
              >
                External port
              </label>
              <div
                class="input-group"
              >
                <input
                  class="form-control"
                  id="3"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-text"
              >
                <small>
                  Enter the external port number to which the rule applies. It can be a single port or a range of ports.
                </small>
              </div>
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="4"
              >
                Internal port
              </label>
              <div
                class="input-group"
              >
                <input
                  class="form-control"
                  id="4"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-text"
              >
                <small>
                  Enter the internal port number to which the rule applies. It can be a single port or a range of ports.
                </small>
              </div>
            </div>
            <div
              class="form-check form-switch mb-3"
            >
              <input
                class="form-check-input"
                id="5"
                role="switch"
                type="checkbox"
              />
              <label
                class="form-check-label"
                for="5"
              >
                Enabled
              </label>
              <div
                class="form-text"
              >
                <small>
                  Enable or disable the port forwarding rule.
                </small>
              </div>
            </div>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-secondary d-inline-flex justify-content-center align-items-center"
              type="button"
            >
              <span>
                Close
              </span>
            </button>
            <button
              class="btn btn-primary d-inline-flex justify-content-center align-items-center"
              disabled=""
              type="button"
            >
              <span>
                Save
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
